src/ctia/task/migration/migrate_es_stores.clj

```diff
@@ -0,0 +1,289 @@
+   restart? :- s/Bool]
+  (let [migration-state (if restart?
+                          (mst/get-migration migration-id @mst/migration-es-conn)
+                          (mst/init-migration migration-id prefix store-keys confirm?))
+        migrations (compose-migrations migrations)
+        batch-size (or batch-size default-batch-size)]
+
+    (log/infof "migrating stores: %s" store-keys)
+    (log/infof "set batch size: %s" batch-size)
+
+    (doseq [entity-type (keys (:stores migration-state))]
+      (log/infof "migrating store: %s" entity-type)
+      (migrate-store migration-state
+                     entity-type
+                     migrations
+                     batch-size
+                     confirm?))
+    (handle-deletes migration-state store-keys confirm?)))
+
+(defn check-store
+  "check a single store"
+  [target-store batch-size]
+  (let [store-schema (type->schema (keyword (:type target-store)))
+        list-coerce (list-coerce-fn store-schema)
+        target-store-size (mst/store-size target-store)]
+    (log/infof "%s - store size: %s records"
+               (:type target-store)
+               target-store-size)
+
+    (loop [offset 0
+           sort-keys nil
+           checked-count 0
+           invalids []]
+      (let [{:keys [data paging]
+             :as batch}
+            (mst/fetch-batch target-store
+                         batch-size
+                         offset
+                         nil
+                         sort-keys)
+            next (:next paging)
+            offset (:offset next 0)
+            search_after (:sort paging)
+            {:keys [errors]} (list-coerce data)
+            checked-count (+ checked-count
+                             (count data))]
+        (if next
+          (recur offset search_after checked-count (concat invalids errors))
+          (do
+            (log/infof "%s - finished checking %s documents"
+                       (:type target-store)
+                       checked-count)
+            (when (seq errors)
+              (log/warnf "%s - errors were detected: %s"
+                         (:type target-store)
+                         (pr-str errors)))))))))
+
+(defn check-store-indexes
+  "check all new es store indexes"
+  [store-keys batch-size prefix]
+  (let [current-stores (mst/stores->maps (select-keys @stores store-keys))
+        target-stores
+        (mst/source-store-maps->target-store-maps current-stores
+                                              prefix)
+        batch-size (or batch-size default-batch-size)]
+
+    (log/infof "checking stores: %s" (keys current-stores))
+    (log/infof "set batch size: %s" batch-size)
+    (doseq [[sk sr] target-stores]
+      (log/infof "checking store: %s" sk)
+      (check-store sr batch-size))))
+
+(defn exit [error?]
+  (if error?
+    (System/exit -1)
+    (System/exit 0)))
+
+(defn run-migration
+  [migration-id prefix migrations store-keys batch-size confirm? restart?]
```

Comment:
IMO the `restart?` option is risky because there is no way to stop the migration process gently. You could register a shutdown hook to the java runtime to stop the process properly.

Reply:
The restart of the migration of a store is mainly based on the search_after field in the source store. In the target store, the ES ids of documents are preserved from source to target store. Thus in the worst case, document will be migrated twice but it will only erased previously written documents.